source/mqtt_client_connection.c

```diff
@@ -53,6 +53,9 @@ struct mqtt_connection_binding {
 
     /* Dependencies that must outlive this */
     PyObject *client;
+
+    /* Reference count - needed to keep the connection alive even if the Python object is destroyed */
+    struct aws_atomic_var ref_count;
```

Comment:
Note: I had to add reference counting in C here, because it is possible for a Python program to be waiting on the disconnect to finish, and then as soon as it finishes it shuts down. Prior to this change, the Python object's cleanup in C was tied 1:1 with the cleanup of the Python object, and so what would happen is the Python object would be cleaned right as the `on_closed` callback is invoked (as `on_closed` happens AFTER the disconnect) and this would lead to unavoidable segfaults!

Using a quick reference count fixes this issue by allowing the C object to live longer than the Python one, allowing the code to tell it to stay alive until the `on_stopped` is called. I also added this reference counting to the connection callback/events as well.

Reply:
After discussion with Vera, we are going to see if it is possible to work around this issue and not use reference counting in C.